runtime/include/wasm_memory.h

```diff
@@ -0,0 +1,302 @@
+ * @return pointer to the string or NULL if max_length is reached without finding null-terminator
+ */
+static inline char *
+wasm_memory_get_string(struct wasm_memory *self, uint32_t offset, uint32_t size)
+{
+	assert(offset + (sizeof(char) * size) <= self->size);
+
+	for (uint32_t i = 0; i < size; i++) {
+		if (self->data[offset + i] == '\0') return (char *)&self->data[offset];
+	}
+	return NULL;
+}
+
+/**
+ * Write a float to WebAssembly linear memory
+ * @param offset an offset into the WebAssembly linear memory
+ * @return float at the offset
+ */
+static inline void
+wasm_memory_set_f32(struct wasm_memory *self, uint32_t offset, float value)
+{
+	assert(offset + sizeof(float) <= self->size);
+	*(float *)&self->data[offset] = value;
+}
+
+/**
+ * Write a double to WebAssembly linear memory
+ * @param offset an offset into the WebAssembly linear memory
+ * @return double at the offset
+ */
+static inline void
+wasm_memory_set_f64(struct wasm_memory *self, uint32_t offset, double value)
+{
+	assert(offset + sizeof(double) <= self->size);
+	*(double *)&self->data[offset] = value;
+}
+
+/**
+ * Write a int8_t to WebAssembly linear memory
+ * @param offset an offset into the WebAssembly linear memory
+ * @return int8_t at the offset
+ */
+static inline void
+wasm_memory_set_i8(struct wasm_memory *self, uint32_t offset, int8_t value)
+{
+	assert(offset + sizeof(int8_t) <= self->size);
+	*(int8_t *)&self->data[offset] = value;
+}
+
+/**
+ * Write a int16_t to WebAssembly linear memory
+ * @param offset an offset into the WebAssembly linear memory
+ * @return int16_t at the offset
+ */
+static inline void
+wasm_memory_set_i16(struct wasm_memory *self, uint32_t offset, int16_t value)
+{
+	assert(offset + sizeof(int16_t) <= self->size);
+	*(int16_t *)&self->data[offset] = value;
+}
+
+/**
+ * Write a int32_t to WebAssembly linear memory
+ * @param offset an offset into the WebAssembly linear memory
+ * @return int32_t at the offset
+ */
+static inline void
+wasm_memory_set_i32(struct wasm_memory *self, uint32_t offset, int32_t value)
+{
+	assert(offset + sizeof(int32_t) <= self->size);
+	*(int32_t *)&self->data[offset] = value;
+}
+
+/**
+ * Write a int64_t to WebAssembly linear memory
+ * @param offset an offset into the WebAssembly linear memory
+ * @return int64_t at the offset
+ */
+static inline void
+wasm_memory_set_i64(struct wasm_memory *self, uint64_t offset, int64_t value)
+{
+	assert(offset + sizeof(int64_t) <= self->size);
+	*(int64_t *)&self->data[offset] = value;
+}
+
+static inline void
+wasm_memory_set_size(struct wasm_memory *self, size_t size)
+{
+	self->size = size;
+}
+
+static inline size_t
+wasm_memory_get_size(struct wasm_memory *self)
+{
+	return self->size;
+}
+
+static inline void
+wasm_memory_initialize_region(struct wasm_memory *self, uint32_t offset, uint32_t region_size, uint8_t region[])
+{
+	assert((size_t)offset + region_size <= self->size);
+	memcpy(&self->data[offset], region, region_size);
+}
```

Comment:
I find all of this a little odd. You're just moving stuff that was already in a function per operation into effectively another set of functions per op by adding a level of indirection. I don't see what you buy by doing this, but am not opposed to it.

Reply:
I would say the main reason is to achieve a layer of abstraction from the ABI symbols that aWsm requires. We can internally organize the state and behavior for wasm components in a cleaner way.